# Часть 1: Задания, решаемые без использования Python

## Задание 1
Исследуйте систему векторов на линейную зависимость или независимость без использования Python (вручную):

а1 = (1, 8, -1)
а2 =  (-2, 3, 3)
a3 = (4, -11, 9)  

Подробно распишите решение, а в итоге сделайте вывод про то, являются ли векторы зависимыми или независимыми

***Определение линейной зависимости***:

Векторы $a_1, a_2, a_3$ линейно зависимы, если существуют такие коэффициенты $w_1, w_2, w_3$, совместно не равные нулю, что выполняется следующее условие:

$
w_1 a_1 + w_2 a_2 + w_3 a_3 = 0.
$

Если таких коэффициентов нет, то векторы линейно независимы.

---
***1. Составим уравнения:***

Подставим векторы $a_1, a_2, a_3$:

$
w_1 (1, 8, -1) + w_2 (-2, 3, 3) + w_3 (4, -11, 9) = (0, 0, 0).
$

Получим систему линейных уравнений:

$
\begin{aligned}
& \quad w_1 - 2w_2 + 4w_3 = 0, \\
& \quad 8w_1 + 3w_2 - 11w_3 = 0, \\
& \quad -w_1 + 3w_2 + 9w_3 = 0.
\end{aligned}
$


***2. Запишем систему в матричном виде:***

$$
\begin{bmatrix}
1 & -2 & 4 \\
8 & 3 & -11 \\
-1 & 3 & 9
\end{bmatrix}
\cdot 
\begin{bmatrix}
w_1 \\
w_2 \\
w_3
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
0
\end{bmatrix}
$$

Будем искать нетривиальные решения данной системы. Чтобы определить, существуют ли такие решения, необходимо найти ранг составленной матрицы матрицы

***3. Находим ранг матрицы***
Найдем определитель:

$
\text{det}
\begin{bmatrix}
1 & -2 & 4 \\
8 & 3 & -11 \\
-1 & 3 & 9
\end{bmatrix}
$


$$
\text{det} =
1 \cdot
\begin{vmatrix}
3 & -11 \\
3 & 9
\end{vmatrix}
-
(-2) \cdot
\begin{vmatrix}
8 & -11 \\
-1 & 9
\end{vmatrix}
+
4 \cdot
\begin{vmatrix}
8 & 3 \\
-1 & 3
\end{vmatrix}
$$

$
\text{det} = 1 \cdot 60 - (-2) \cdot 61 + 4 \cdot 27 = 60 + 122 + 108 = 290
$


Так как определитель матрицы $ \neq 0 $, ранг матрицы равен 3 (максимальный). Следовательно, система уравнений имеет только тривиальное решение $w_1 = w_2 = w_3 = 0$


***Вывод:***

Векторы $a_1 = (1, 8, -1)$, $a_2 = (-2, 3, 3)$, $a_3 = (4, -11, 9)$ линейно независимы

## Задание 2

При каких значениях параметра a система является линейно зависимой?  
Найдите решение без использования Python (вручную)

a1 = (1; 0; 2; -3),  
a2 = (2; 2; -4; a),  
a3 = (3; 1; 2; -5)

Для того чтобы система векторов была линейно зависимой, должно существовать нетривиальное решение уравнения:

$
w_1 \mathbf{a}_1 + w_2 \mathbf{a}_2 + w_3 \mathbf{a}_3 = \mathbf{0}
$

где $\mathbf{0}$ — нуль-вектор, а $ w_1, w_2 и w_3 $ совместно не равны нулю. 

Составим систему уравнений:

$
\begin{aligned}
& \quad w_1 + 2w_2 + 3w_3 = 0, \\
& \quad 0 w_1 + 2 w_2 + w_3 = 0, \\
& \quad 2 w_1 - 4 w_2 + 2 w_3 = 0 \\
& \quad -3 w_1 + a c_2 - 5 c_3 = 0.
\end{aligned}
$

Из уравнения 2 определим $ w_3 = -2 w_2 $

Подставим выраженное $ w_3 $ в уравнение 1:

(1) $ w_1 + 2w_2 + 3(-2 w_2) = 0 $

Тогда:

$ w_1 = 4 w_2 $

Подставим переопределённые $ w_1, w_3 $ в уравнение 3:

(3) $ 2(4 w_2) - 4 w_2 = 2(-2 w_w) = 0 $

Тогда, очевидно:

(3) $ 0 = 0 $

То, есть решением уравнения является множество действительных чисел

Подставим переопределённые $ w_1, w_3 $ в уравнение 4:

(4) $ -3(4 w_2) + a w_2 - 5 (-2 w_2) = 0 $

Тогда:

(4) $ w_2 (a - 2) = 0 $

В таком случае, решением уравнения является либо

$ w_2 = 0 $ либо $ a = 2 $

Поскольку при $ w_2 = 0 => w_1 = w_3 = 0 $, что противоречит критерию, подходит только решение $ a = 2 $. В таком случае система векторов является линейно зависимой.

***Ответ:*** $ a = 2 $

***Проверка:***

$
A = \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2 \\
-3 & 2 & -5
\end{bmatrix}
$

$ III = III - 2 I $

$
A = \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
0 & -8 & -4 \\
-3 & 2 & -5
\end{bmatrix}
$

$ IV = IV + 3I $

$
A = \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
0 & -8 & -4 \\
0 & 8 & 4
\end{bmatrix}
$

$ IV = IV + III $

$
A = \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
0 & -8 & -4 \\
0 & 0 & 0
\end{bmatrix}
$

Таким образом, $ 4 a_1 + a_2 - 2 a_3 = 0 $

Система векторов является линейно зависимой при $ a = 2 $


## Задание 3
Выразите определитель матрицы. Найдите решение без использования Python (вручную)

\begin{vmatrix}
1 & -1 & a & -1 \\
-1 & -2 & b & -1 \\
-2 & 0 & c & 1 \\
0 & 1 & d & 0
\end{vmatrix}

Чтобы найти определитель этой матрицы $ \det(A) $, мы можем разложить его по элементам 4-ой строки
$
\det(A) = -0 \cdot \det(A_{41}) + 1 \cdot \det(A_{42}) - d \cdot \det(A_{43}) + 0 \cdot \det(A_{44})
$

где $A_{ij}$ — это дополнение элемента $a_{ij}$, то есть определитель матрицы, полученной из $A$ исключением $i$-й строки и $j$-го столбца.

Теперь вычислим каждый из этих определителей меньшего порядка $3 \times 3$:

$$ 

-0 \cdot 
\begin{vmatrix}
1 & a & -1\\
-2 & b & -1\\
0 & c & 1
\end{vmatrix}
+ 1 \cdot
\begin{vmatrix}
1 & a & -1\\
-1 & b & -1\\
-2 & c & 1
\end{vmatrix}
-d \cdot 
\begin{vmatrix}
1 & -1 & -1\\
-1 & -2 & -1\\
-2 & 0 & 1
\end{vmatrix}
+0 \cdot
\begin{vmatrix}
1 & -1 & a\\
-1 & -2 & b\\
-2 & 0 & c
\end{vmatrix} 
= 0 - b + 3a + 2c - d -1 + 0 = -b+3a+2c+d
$$

***Ответ:*** $-b+3a+2c+d$


## Задание 4
Найдите значения параметров a, b и c, при которых матрицы A и B являются обратными. Найдите решение без использования Python (вручную).

$A=\left(\begin{array}{ccc}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2 c
\end{array}\right), B=\left(\begin{array}{rrr}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{array}\right)$

Для того чтобы матрицы $ A $ и $ B $ были обратными, произведение $ AB $ и $ BA $ должно быть равно единичной матрице $ I $.

Вычислим произведение $ AB $:

$
AB = \begin{pmatrix}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2c
\end{pmatrix} 
\begin{pmatrix}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{pmatrix}
$

$
AB = \begin{pmatrix}
-4a + 5 & 3a - 3 & 5a - 5 \\
0 & 2b - 17 & 3b - 27 \\
-40 - 10c & -8 - 2c & -7 - 2c
\end{pmatrix} 
$

Для того чтобы $ AB = I $, матрица произведения должна быть равна единичной матрице:

$
\begin{pmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1
\end{pmatrix}
$

Составим уравнения для нахождения коэффициентов:

1. $-4a + 5 = 1$
2. $3a - 3 = 0$
3. $5a - 5 = 0$
4. $2b - 17 = 1$
5. $3b - 27 = 0$
6. $-40 - 10c = 0$
7. $-8 - 2c = 0$
8. $-7 - 2c = 1$

Решив эти уравнения, получим:

1. $ a = 1 $

2. $ b = 9 $

3. $ c = -4 $

***Ответ:*** $ a = 1, b = 9, c = -4 $


## Задание 5
Найдите собственные векторы и собственные значения матрицы. Найдите решение без использования Python (вручную).

$\left(\begin{array}{rrr}4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4\end{array}\right)$

Для нахождения собственных значений и собственных векторов матрицы, нужно сначала найти собственные значения $\lambda$, решив характеристическое уравнение $\det(A - \lambda I) = 0$.

1. **Найдем характеристический многочлен:**

$
A - \lambda I = \begin{pmatrix} 4 - \lambda & 0 & -1 \\ 2 & 2 - \lambda & 1 \\ -1 & 0 & 4 - \lambda \end{pmatrix}
$

Найдем определитель характеристического многочлена:

$
\det(A - \lambda I) = \begin{vmatrix} 4 - \lambda & 0 & -1 \\ 2 & 2 - \lambda & 1 \\ -1 & 0 & 4 - \lambda \end{vmatrix} = - \lambda^3 +10\lambda^2-31\lambda+30 = 
-(\lambda-2)(\lambda^2-8\lambda+15)=-(\lambda-2)(\lambda-3)(\lambda-5)=0
$




Теперь найдем собственные векторы для данных собственных значений.

***$\lambda = 2$:***

$(A - 2I)\mathbf{v} = \mathbf{0}$:

$
A - 2I = \begin{pmatrix} 2 & 0 & -1 \\ 2 & 0 & 1 \\ -1 & 0 & 2 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
2x - z = 0 \\
2x + z = 0 \\
-x + 2z = 0
\end{cases}
$

Решение: $x = 0$, $z = 0$, $y$ свободный параметр, следовательно, $\mathbf{v}_1 = \begin{pmatrix} 0 \\ 1 \\ 0 \end{pmatrix}$.

***$\lambda = 3$:***

$(A - 3I)\mathbf{v} = \mathbf{0}$:

$
A - 3I = \begin{pmatrix} 1 & 0 & -1 \\ 2 & -1 & 1 \\ -1 & 0 & 1 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
x - z = 0 \\
2x - y + z = 0 \\
-x + z = 0
\end{cases}
$

Решение: $x = z$, $y = 3z$, следовательно, $\mathbf{v}_2 = \begin{pmatrix} 1 \\ 3 \\ 1 \end{pmatrix}$.

***$\lambda = 5$:***

$(A - 5I)\mathbf{v} = \mathbf{0}$:

$
A - 5I = \begin{pmatrix} -1 & 0 & -1 \\ 2 & -3 & 1 \\ -1 & 0 & -1 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
-x - z = 0 \\
2x - 3y + z = 0 \\
-x - z = 0
\end{cases}
$

Решение: $x = -z$, $y = -\frac{1}{3}z$, следовательно, $\mathbf{v}_3 = \begin{pmatrix} -1 \\ -\frac{1}{3} \\ 1 \end{pmatrix}$.

***Ответ:***

1. $\lambda_1 = 2$, $\mathbf{v}_1 = \begin{pmatrix} 0 \\ 1 \\ 0 \end{pmatrix}$

2. $\lambda_2 = 3$, $\mathbf{v}_2 = \begin{pmatrix} 1 \\ 3 \\ 1 \end{pmatrix}$

3. $\lambda_3 = 5$, $\mathbf{v}_3 = \begin{pmatrix} -1 \\ -\frac{1}{3}\\ 1 \end{pmatrix}$

## Задание 6
Найдите сумму сингулярных чисел для матрицы:

$A=\left(\begin{array}{ccc}-2 & 0 & 1 \\ 0 & -2 & 1 \\ 1 & 1 & -1\end{array}\right)$

Найдите решение без использования Python (вручную).

***Найдём сумму сингулярных чисел для матрицы $A$***

Сингулярные числа матрицы $ A $ — это квадратные корни собственных значений матрицы $ A^T A $ (или $ AA^T $)

***Найдём $ A^T A $***

$
A^T A =
\begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}
\begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}
$

$
A^T A =
\begin{pmatrix}
5 & 1 & -3 \\
1 & 5 & -3 \\
-3 & -3 & 3
\end{pmatrix}
$

***Найдём собственные значения $A^T A$***

Решаем характеристическое уравнение:

$
\det(A^T A - \lambda I) = 0.
$

$
A^T A - \lambda I =
\begin{bmatrix}
5 - \lambda & 1 & -3 \\
1 & 5 - \lambda & -3 \\
-3 & -3 & 3 - \lambda
\end{bmatrix} = -\lambda^3 + 13\lambda^2-36\lambda=-\lambda(\lambda^2-13\lambda+36) = -\lambda(\lambda-4)(\lambda-9)=0
$

$
\sigma_1 = \sqrt{0} = 0,
$
$
\sigma_2 = \sqrt{4} = 2,
$
$
\sigma_3 = \sqrt{9} = 3.
$

$
\sigma_1 + \sigma_2 + \sigma_3 = 0 + 2 + 3 = 5.
$

***Ответ:***
Сумма сингулярных чисел матрицы $A$ равна $5$.

# Часть 2: Задания, решаемые с помощью Python

Даны данные об аренде велосипедов.

Описание переменных:

- season: 1 - весна, 2 - лето, 3 - осень, 4 - зима
- yr: 0 - 2011, 1 - 2012
- mnth: месяц, от 1 до 12
- holiday: 0 - нет праздника, 1 - есть праздник
- weekday: день недели, от 0 до 6
- workingday: 0 - нерабочий день, 1 - рабочий день
- weathersit: оценка благоприятности погоды от 1 (чистый, ясный день) до 4 (ливень, туман)
- temp: температура в Цельсиях
- atemp: температура по ощущениям в Цельсиях
- hum: влажность
- windspeed(mph): скорость ветра в милях в час
- windspeed(ms): скорость ветра в метрах в секунду
- cnt: количество арендованных велосипедов (это целевой признак, его мы будем предсказывать)

## Задание 7
Загрузите данные. Изучите, есть ли в них пропущенные значения? Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [38]:
import pandas as pd

In [39]:
data = pd.read_csv(
    "data/asset_v1_SkillFactory+MIFIML_1sem+2024+type@asset+block@bikes_rent.xls"
)
data.head()

season  yr  mnth  holiday  weekday  workingday  weathersit       temp  \
0       1   0     1        0        6           0           2  14.110847   
1       1   0     1        0        0           0           2  14.902598   
2       1   0     1        0        1           1           1   8.050924   
3       1   0     1        0        2           1           1   8.200000   
4       1   0     1        0        3           1           1   9.305237   

      atemp      hum  windspeed(mph)  windspeed(ms)   cnt  
0  18.18125  80.5833       10.749882       4.805490   985  
1  17.68695  69.6087       16.652113       7.443949   801  
2   9.47025  43.7273       16.636703       7.437060  1349  
3  10.60610  59.0435       10.739832       4.800998  1562  
4  11.46350  43.6957       12.522300       5.597810  1600

In [40]:
data.isnull().sum()

season            0
yr                0
mnth              0
holiday           0
weekday           0
workingday        0
weathersit        0
temp              0
atemp             0
hum               0
windspeed(mph)    0
windspeed(ms)     0
cnt               0
dtype: int64

В данном датасете не встречается пустых значений

## Задание 8
Исследуем данные на наличие линейной зависимости между столбцами. Найдите ранг матрицы данных. Сравните его с количеством столбцов, сделайте вывод. Если линейная зависимость есть - напишите, между какими столбцами и удалите один из них.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [41]:
import numpy as np

In [42]:
# рассчитаем корреляцию между числовыми столбцами
display("Корреляционная матрица:", data.corr())

'Корреляционная матрица:'

season        yr      mnth   holiday   weekday  workingday  \
season          1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr             -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth            0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday        -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday        -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday      0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit      0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp            0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp           0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum             0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(mph) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
windspeed(ms)  -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt             0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

                weathersit      temp     atemp       hum  windspeed(mph)  \
season            0.019211  0.334315  0.342876  0.205445       -0.229046   
yr               -0.048727  0.047604  0.046106 -0.110651       -0.011817   
mnth              0.043528  0.220205  0.227459  0.222204       -0.207502   
holiday          -0.034627 -0.028556 -0.032507 -0.015937        0.006292   
weekday           0.031087 -0.000170 -0.007537 -0.052232        0.014282   
workingday        0.061200  0.052660  0.052182  0.024327       -0.018796   
weathersit        1.000000 -0.120602 -0.121583  0.591045        0.039511   
temp             -0.120602  1.000000  0.991702  0.126963       -0.157944   
atemp            -0.121583  0.991702  1.000000  0.139988       -0.183643   
hum               0.591045  0.126963  0.139988  1.000000       -0.248489   
windspeed(mph)    0.039511 -0.157944 -0.183643 -0.248489        1.000000   
windspeed(ms)     0.039511 -0.157944 -0.183643 -0.248489        1.000000   
cnt              -0.297391  0.627494  0.631066 -0.100659       -0.234545   

                windspeed(ms)       cnt  
season              -0.229046  0.406100  
yr                  -0.011817  0.566710  
mnth                -0.207502  0.279977  
holiday              0.006292 -0.068348  
weekday              0.014282  0.067443  
workingday          -0.018796  0.061156  
weathersit           0.039511 -0.297391  
temp                -0.157944  0.627494  
atemp               -0.183643  0.631066  
hum                 -0.248489 -0.100659  
windspeed(mph)       1.000000 -0.234545  
windspeed(ms)        1.000000 -0.234545  
cnt                 -0.234545  1.000000

In [43]:
# найдем ранг матрицы
matrix_rank = np.linalg.matrix_rank(data)
print("Ранг матрицы:", matrix_rank)

Ранг матрицы: 12


In [44]:
# cравним ранг с количеством столбцов
num_columns = data.shape[1]
print("Количество столбцов:", num_columns)

Количество столбцов: 13


In [45]:
# вывод по линейной зависимости
if matrix_rank < num_columns:

    display("Есть линейная зависимость между некоторыми столбцами")

    # Находим пары столбцов с высокой корреляцией (выше 0.8)
    cor_mx = data.corr()
    high_correlation = cor_mx[
        (cor_mx > 0.8) & (cor_mx < 1.0)
    ]
    display("Столбцы с высокой корреляцией:")
    display(high_correlation)

else:
    display("Нет линейной зависимости между столбцами")

'Есть линейная зависимость между некоторыми столбцами'

'Столбцы с высокой корреляцией:'

season  yr     mnth  holiday  weekday  workingday  \
season              NaN NaN  0.83144      NaN      NaN         NaN   
yr                  NaN NaN      NaN      NaN      NaN         NaN   
mnth            0.83144 NaN      NaN      NaN      NaN         NaN   
holiday             NaN NaN      NaN      NaN      NaN         NaN   
weekday             NaN NaN      NaN      NaN      NaN         NaN   
workingday          NaN NaN      NaN      NaN      NaN         NaN   
weathersit          NaN NaN      NaN      NaN      NaN         NaN   
temp                NaN NaN      NaN      NaN      NaN         NaN   
atemp               NaN NaN      NaN      NaN      NaN         NaN   
hum                 NaN NaN      NaN      NaN      NaN         NaN   
windspeed(mph)      NaN NaN      NaN      NaN      NaN         NaN   
windspeed(ms)       NaN NaN      NaN      NaN      NaN         NaN   
cnt                 NaN NaN      NaN      NaN      NaN         NaN   

                weathersit      temp     atemp  hum  windspeed(mph)  \
season                 NaN       NaN       NaN  NaN             NaN   
yr                     NaN       NaN       NaN  NaN             NaN   
mnth                   NaN       NaN       NaN  NaN             NaN   
holiday                NaN       NaN       NaN  NaN             NaN   
weekday                NaN       NaN       NaN  NaN             NaN   
workingday             NaN       NaN       NaN  NaN             NaN   
weathersit             NaN       NaN       NaN  NaN             NaN   
temp                   NaN       NaN  0.991702  NaN             NaN   
atemp                  NaN  0.991702       NaN  NaN             NaN   
hum                    NaN       NaN       NaN  NaN             NaN   
windspeed(mph)         NaN       NaN       NaN  NaN             NaN   
windspeed(ms)          NaN       NaN       NaN  NaN             1.0   
cnt                    NaN       NaN       NaN  NaN             NaN   

                windspeed(ms)  cnt  
season                    NaN  NaN  
yr                        NaN  NaN  
mnth                      NaN  NaN  
holiday                   NaN  NaN  
weekday                   NaN  NaN  
workingday                NaN  NaN  
weathersit                NaN  NaN  
temp                      NaN  NaN  
atemp                     NaN  NaN  
hum                       NaN  NaN  
windspeed(mph)            1.0  NaN  
windspeed(ms)             NaN  NaN  
cnt                       NaN  NaN

На основе полученной корреляционной матрицы видно, что значения корреляции между некоторыми столбцами очень высоки, что соответствует физическому смыслу признаков.

- temp и atemp: корреляция 0.991702 — зависимость между температурой и температурой по ощущениям.
- windspeed(mph) и windspeed(ms): корреляция 1.0 — линейная зависимость, так как скорость ветра измеряется в двух единицах. 

Удалим столбцы windspeed(mph) и atemp, поскольку мультиколлинеарность признаков близка к линейной зависимости

In [46]:
del data["windspeed(mph)"]
del data["atemp"]

## Задание 9
Выведите корреляционную матрицу оставшихся переменных.

Какая переменная сильнее всего коррелирует с целевой?
Есть ли мультиколлинеарность? Между какими переменными? Если есть корреляция выше 0.9 - оставьте только 1 признак. Объясните выбор
Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [47]:
data.corr()

season        yr      mnth   holiday   weekday  workingday  \
season         1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr            -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth           0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday       -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday       -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday     0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit     0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp           0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
hum            0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(ms) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt            0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

               weathersit      temp       hum  windspeed(ms)       cnt  
season           0.019211  0.334315  0.205445      -0.229046  0.406100  
yr              -0.048727  0.047604 -0.110651      -0.011817  0.566710  
mnth             0.043528  0.220205  0.222204      -0.207502  0.279977  
holiday         -0.034627 -0.028556 -0.015937       0.006292 -0.068348  
weekday          0.031087 -0.000170 -0.052232       0.014282  0.067443  
workingday       0.061200  0.052660  0.024327      -0.018796  0.061156  
weathersit       1.000000 -0.120602  0.591045       0.039511 -0.297391  
temp            -0.120602  1.000000  0.126963      -0.157944  0.627494  
hum              0.591045  0.126963  1.000000      -0.248489 -0.100659  
windspeed(ms)    0.039511 -0.157944 -0.248489       1.000000 -0.234545  
cnt             -0.297391  0.627494 -0.100659      -0.234545  1.000000

In [48]:
data.corr()["cnt"].to_frame().sort_values(by="cnt", ascending=False)

cnt
cnt            1.000000
temp           0.627494
yr             0.566710
season         0.406100
mnth           0.279977
weekday        0.067443
workingday     0.061156
holiday       -0.068348
hum           -0.100659
windspeed(ms) -0.234545
weathersit    -0.297391

- Самая сильная корреляция с целевой переменной cnt: это признак temp (0.627494), отражающий температуру.
- Мультиколлинеарность признаков была устранена на предыдущем шаге.

## Задание 10
Разделите данные на предикторы и целевую переменную. Обучите модель с помощью классической линейной регрессии. Линейную регрессию необходимо реализовать без готовых библиотек.

Оцените качество модели с помощью хотя бы двух метрик качества.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [49]:
X = data.drop(columns=["cnt"])  # все переменные, кроме целевой переменной
y = data["cnt"]  # целевая переменная

In [50]:
# Формируем bias
X = np.c_[np.ones(X.shape[0]), X]

X = np.array(X)
y = np.array(y)

# находим коэффициенты линейной регрессии
w_hat = np.linalg.inv(X.T @ X) @ (X.T @ y)

display(w_hat)

array([1626.11274728,  515.23642918, 2040.88251485,  -39.80545835,
       -536.84058293,   67.21118086,  119.21163986, -622.29155022,
        125.71269207,   -9.60150621,  -91.16358693])

In [51]:
# получаем прогнозы
y_pred = X @ w_hat
display(y_pred)
display(y)

array([1822.33347068, 1383.44091148, 1579.94131911, 1759.14753564,
       2040.02281018, 2178.40349037, 1385.62892608,  867.70541323,
        787.10795657, 1369.81469465,  989.64183183, 1282.95574723,
       1444.22584742, 1901.7553758 , 1552.69618752, 1694.33993648,
        247.7834557 ,  999.25587727, 1402.75991529, 1541.92445562,
       1446.02457784, 1333.67896076,  884.29665058, 1263.19954345,
       1315.67961367,   46.09193748, 1901.45458224, 1262.89650618,
       1872.7611401 , 1699.85453023,  884.63195115, 1119.4948616 ,
       1011.90740185, 1612.36079096, 1450.57884112, 1091.11757572,
       1979.83345969, 2193.20627926, 1326.54405383,  842.02472924,
       1545.56640275, 2086.75761541, 1911.76724076, 1919.38517999,
       2262.94832161, 1967.15551896, 2306.58016211, 2960.7246651 ,
       3364.06499764, 2334.09028404, 1908.41721032,  489.52940133,
       1543.24143074, 2240.03009351, 1375.50618747, 1520.35959351,
       2272.88161993, 2214.92534548, 1470.34549162, 1923.12296

array([ 985,  801, 1349, 1562, 1600, 1606, 1510,  959,  822, 1321, 1263,
       1162, 1406, 1421, 1248, 1204, 1000,  683, 1650, 1927, 1543,  981,
        986, 1416, 1985,  506,  431, 1167, 1098, 1096, 1501, 1360, 1526,
       1550, 1708, 1005, 1623, 1712, 1530, 1605, 1538, 1746, 1472, 1589,
       1913, 1815, 2115, 2475, 2927, 1635, 1812, 1107, 1450, 1917, 1807,
       1461, 1969, 2402, 1446, 1851, 2134, 1685, 1944, 2077,  605, 1872,
       2133, 1891,  623, 1977, 2132, 2417, 2046, 2056, 2192, 2744, 3239,
       3117, 2471, 2077, 2703, 2121, 1865, 2210, 2496, 1693, 2028, 2425,
       1536, 1685, 2227, 2252, 3249, 3115, 1795, 2808, 3141, 1471, 2455,
       2895, 3348, 2034, 2162, 3267, 3126,  795, 3744, 3429, 3204, 3944,
       4189, 1683, 4036, 4191, 4073, 4400, 3872, 4058, 4595, 5312, 3351,
       4401, 4451, 2633, 4433, 4608, 4714, 4333, 4362, 4803, 4182, 4864,
       4105, 3409, 4553, 3958, 4123, 3855, 4575, 4917, 5805, 4660, 4274,
       4492, 4978, 4677, 4679, 4758, 4788, 4098, 39

In [52]:
# Средняя абсолютная ошибка
mae = np.mean((y - y_pred))

# Средняя абсолютная ошибка в процентах
mape = np.mean((abs(y - y_pred)/y)*100)

print(f"MAE: {mae}")
print(f"MAPE: {mape}")

MAE: -5.763533866611372e-11
MAPE: 44.87517666522052


***Вывод:***

Как видно, полученные результаты не являются точными, поскольку средняя относительная ошибка составляет примерно 45%. Скорее всего, нормализация признаков смогла бы улучшить качество результатов работы модели.